### ISSUE TITLE
Saving model containing KerasLayer with trainable=True more than doubles size of SavedModel

### ISSUE THREAD

**Issue by dwyatte** [tensorflow/hub#736]
When embedding a TensorFlow Hub KerasLayer in a model and saving, the size of the SavedModel increases when the KerasLayer is trainable. For example when instantiating https://tfhub.dev/google/universal-sentence-encoder/4 as trainable, the size of the SavedModel increases almost 3x (from ~1 GB which is expected for the model to almost 3 GB). Furthermore, this larger SavedModel can be reloaded and resaved, which decreases its size back down around the expected size with the same weights.

I suspect something is going on with the graph tracing when saving a KerasLayer and could maybe justify a 2x increase (to instantiate the TensorFlow Hub model with its default weights and assign its trained weights), but not a 3x increase.

Example:

```
import subprocess
import tensorflow as tf
import tensorflow_hub as hub


class MyModel(tf.keras.Model):

    def __init__(self, trainable_encoder=False):
        super().__init__()
        self.encoder = hub.KerasLayer("https://tfhub.dev/google/universal-sentence-encoder/4", trainable=trainable_encoder)
        self.dense = tf.keras.layers.Dense(1, activation="sigmoid")

    def call(self, inputs, training=None):
        return self.dense(self.encoder(tf.squeeze(inputs, axis=1)))


# untrainable KerasLayer, ~1 GB
untrainable_encoder = MyModel(trainable_encoder=False)
untrainable_encoder.compile(loss="binary_crossentropy", optimizer="adam")
untrainable_encoder.fit(["a sentence"], [1], epochs=1)
untrainable_encoder.save("untrainable_encoder")
untrainable_encoder_size = subprocess.check_output(['du','-sh', "untrainable_encoder"]).split()[0].decode('utf-8')
print("\n untrainable_encoder_size {} \n".format(untrainable_encoder_size))

# trainable KerasLayer, ~2.9 GB
trainable_encoder = MyModel(trainable_encoder=True)
trainable_encoder.compile(loss="binary_crossentropy", optimizer="adam")
trainable_encoder.fit(["a sentence"], [1], epochs=1)
trainable_encoder.save("trainable_encoder")
trainable_encoder_size = subprocess.check_output(['du','-sh', "trainable_encoder"]).split()[0].decode('utf-8')
print("\n trainable_encoder_size {} \n".format(trainable_encoder_size))

# reload SavedModel with trainable KerasLayer, ~1 GB
reloaded_trainable_encoder = tf.keras.models.load_model("trainable_encoder")
reloaded_trainable_encoder.save("reloaded_trainable_encoder")
reloaded_trainable_encoder_size = subprocess.check_output(['du','-sh', "reloaded_trainable_encoder"]).split()[0].decode('utf-8')
print("\n reloaded_trainable_encoder_size {} \n".format(reloaded_trainable_encoder_size))

# weights are the same
assert all([(x.numpy() == y.numpy()).all() for x, y in zip(reloaded_trainable_encoder.weights, trainable_encoder.weights)])
twice_reloaded_trainable_encoder = tf.keras.models.load_model("reloaded_trainable_encoder")
assert all([(x.numpy() == y.numpy()).all() for x, y in zip(twice_reloaded_trainable_encoder.weights, reloaded_trainable_encoder.weights)])
```

Output:

```
2021-02-17 08:54:56.873001: I tensorflow/compiler/jit/xla_cpu_device.cc:41] Not creating XLA devices, tf_xla_enable_xla_devices not set
2021-02-17 08:54:56.873171: I tensorflow/core/platform/cpu_feature_guard.cc:142] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2021-02-17 08:54:59.496029: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
1/1 [==============================] - 2s 2s/step - loss: 0.6452
2021-02-17 08:55:06.479114: W tensorflow/python/util/util.cc:348] Sets are not currently considered sequences, but this may change in the future, so consider avoiding using them.
INFO:tensorflow:Assets written to: untrainable_encoder/assets
INFO:tensorflow:Assets written to: untrainable_encoder/assets

 untrainable_encoder_size 999M 

1/1 [==============================] - 6s 6s/step - loss: 0.7075
INFO:tensorflow:Assets written to: trainable_encoder/assets
INFO:tensorflow:Assets written to: trainable_encoder/assets

 trainable_encoder_size 2.9G 

WARNING:absl:Found untraced functions such as dense_1_layer_call_and_return_conditional_losses, dense_1_layer_call_fn, dense_1_layer_call_fn, dense_1_layer_call_and_return_conditional_losses, dense_1_layer_call_and_return_conditional_losses while saving (showing 5 of 5). These functions will not be directly callable after loading.
WARNING:absl:Found untraced functions such as dense_1_layer_call_and_return_conditional_losses, dense_1_layer_call_fn, dense_1_layer_call_fn, dense_1_layer_call_and_return_conditional_losses, dense_1_layer_call_and_return_conditional_losses while saving (showing 5 of 5). These functions will not be directly callable after loading.
INFO:tensorflow:Assets written to: reloaded_trainable_encoder/assets
INFO:tensorflow:Assets written to: reloaded_trainable_encoder/assets

 reloaded_trainable_encoder_size 1002M 
```

**Comment by Rishabhc711** [comment]
Hi!
I am just a student but after going through the problem, can it be due to the fact that despite of the weights being the same, there are other variables involved in the trainable_encoder? Maybe biases or other variables?

**Comment by dwyatte** [comment]
@Rishabhc711 Good hypothesis. Here's some code to test:

```
untrainable_encoder_trainable_weights = sum([np.prod(w.shape.as_list()) for w in untrainable_encoder.trainable_weights])
untrainable_encoder_non_trainable_weights = sum([np.prod(w.shape.as_list()) for w in untrainable_encoder.non_trainable_weights])
print("\n untrainable_encoder")
print("trainable weights: ", untrainable_encoder_trainable_weights)
print("untrainable weights: ", untrainable_encoder_non_trainable_weights)

trainable_encoder_trainable_weights = sum([np.prod(w.shape.as_list()) for w in trainable_encoder.trainable_weights])
trainable_encoder_non_trainable_weights = sum([np.prod(w.shape.as_list()) for w in trainable_encoder.non_trainable_weights])
print("\n trainable_encoder")
print("trainable weights: ", trainable_encoder_trainable_weights)
print("untrainable weights: ", trainable_encoder_non_trainable_weights)

reloaded_trainable_encoder_trainable_weights = sum([np.prod(w.shape.as_list()) for w in reloaded_trainable_encoder.trainable_weights])
reloaded_trainable_encoder_non_trainable_weights = sum([np.prod(w.shape.as_list()) for w in reloaded_trainable_encoder.non_trainable_weights])
print("\n reloaded_trainable_encoder")
print("trainable weights: ", reloaded_trainable_encoder_trainable_weights)
print("untrainable weights: ", reloaded_trainable_encoder_non_trainable_weights)
```

Output:

```
 untrainable_encoder
trainable weights:  513
untrainable weights:  256797824

 trainable_encoder
trainable weights:  256798337
untrainable weights:  0

 reloaded_trainable_encoder
trainable weights:  256798337
untrainable weights:  0
```

So `untrainable_encoder` has 513 trainable parameters as expected (512 weights + 1 bias) and 256797824 is the number of parameters in of  https://tfhub.dev/google/universal-sentence-encoder/4 for a total of 256798337 parameters. Therefore I'd expect the size of all three models to have an upper bound of roughly 1.02 GB (256798337 parameters * 32 bits of precision) which is only true of `untrainable_encoder` where the `KerasLayer`'s `trainable` attribute is False and `reloaded_trainable_encoder` which has been through a serialization/deserialization roundtrip.

**Comment by maringeo** [comment]
The 3x increase in memory consumption is because of `optimizer="adam"`. 

The Adam optimizer introduces 2 moving-average accumulators (slot variables) for each trainable variable. These two accumulators store the first and second moment of the gradients and explain the 3x increase in memory consumption. 

One way is to decrease the memory usage is to try an optimizer that is less memory-intensive.
